lib/pull_image_to_stream.js

```diff
@@ -14,17 +16,48 @@ export const IMAGE_ERROR = 'Pulling docker image "%s" has failed this may indica
                   'Error with the registry used or an authentication error ' +
                   'in the worker try pulling the image locally. \n Error %s';
 
+export async function pullImageStreamTo(docker, image, stream, options) {
+  // Settings for exponential backoff for retrying image pulls.
+  // Last attempt will be in a range between 6 and 10 minutes which is acceptable
+  // for an image pull.
+  let maxAttempts = 5;
+  let attempts = 0;
+  let delayFactor = 15 * 1000;
+  let randomizationFactor = 0.25;
+  let error;
 
-export function pullImageStreamTo(docker, image, stream, options) {
-  return new Promise(function(accept, reject) {
-    debug('pull image', image);
-    var downloadProgress =
+  while (attempts++ < maxAttempts) {
+    error = undefined;
+
+    debug(`pull image. Image: ${image} Attempts: ${attempts}`);
+    let downloadProgress =
       dockerUtils.pullImageIfMissing(docker, image, options);
 
-    downloadProgress.pipe(stream, { end: false });
-    downloadProgress.once('error', reject);
-    downloadProgress.once('end', accept);
-  });
+    downloadProgress.pipe(stream, {end: false});
+    downloadProgress.once('error', err => { error = err; });
+    await Promise.race([
+      waitForEvent(downloadProgress, 'error'),
+      waitForEvent(downloadProgress, 'end')
+    ]);
+
+    if (!error) {
+      return;
+    }
+
+    if (attempts === maxAttempts) {
+      throw new Error(error);
+    }
+
+    let delay = Math.pow(2, attempts - 1) * delayFactor;
+    delay = delay * (Math.random() * 2 * randomizationFactor + 1 - randomizationFactor);
+    debug(
+      `pull image failed ` +
+      `Next Attempt in: ${delay} ms` +
+      `Image: ${image} ` +
+      `Error: ${error} `
+    );
```

Comment:
String templating compiles to `"mystring " + myvar + " another string"`, this expression is given as argument to `debug()`, hence, following common function call semantics the expression `"mystring " + myvar + " another string"` will be evaluated before the function `debug` is invoked with it as argument.

If you call `debug("my string %s another string", myvar)` there is no expressions to evaluate prior to invocation of `debug`. In the case where `DEBUG` doesn't activate the `debug` function for this file, the result is that `debug = function() {}`. Hence, the string interpolation doesn't take place, if `DEBUG` disables the debug logging, and the call to `debug` has no overhead (a good JS compiler will inline it and dead code elimination may remove it, but even if not overhead is very small). 

On the other hand, if you call `debug("mystring " + myvar + " another string")` where `debug` is an empty function. The JS _may_ be able to recognize that the function `debug` is empty and the expression `"mystring " + myvar + " another string"` has no side effects and then remove it. But this is very very hard. I seriously doubt JS compilers are that great.

However, when using `debug` for string interpolation, even a naive non-optimizing JS interpreter would be able to benefit. String operations are fairly heavy, so being able to disable them when not needed is great. It means that we don't have to feel bad about adding an excessive amount of `debug` statements (as long as we disable them at run time).

Note, using `debug` for string interpolation starts becoming increasingly important if dumping json, as `debug("%j", data)` won't call `JSON.stringify` unless `debug` is activated.